File: code/Api/controllers/permissionController.js
const pool = require('../config/database');
const logger = require('../utils/logger');

class PermissionController {
  async getPermissionTree(req, res) {
    try {
      const [permissions] = await pool.query(`
        SELECT id, permission_name, permission_code, permission_type, parent_id, path, component, icon, sort_order, description, status
        FROM permissions
        ORDER BY sort_order ASC, id ASC
      `);

      const buildTree = (parentId = 0) => {
        return permissions
          .filter(p => p.parent_id === parentId)
          .map(p => ({
            ...p,
            children: buildTree(p.id)
          }));
      };

      const tree = buildTree(0);

      res.json({
        success: true,
        data: tree
      });
    } catch (error) {
      logger.error('获取权限树失败:', error);
      res.status(500).json({
        success: false,
        message: '获取权限树失败'
      });
    }
  }

  async getAllPermissions(req, res) {
    try {
      const [permissions] = await pool.query(`
        SELECT id, permission_name, permission_code, permission_type, parent_id, path, component, icon, sort_order, description, status
        FROM permissions
        WHERE status = 1
        ORDER BY sort_order ASC, id ASC
      `);

      res.json({
        success: true,
        data: permissions
      });
    } catch (error) {
      logger.error('获取所有权限失败:', error);
      res.status(500).json({
        success: false,
        message: '获取所有权限失败'
      });
    }
  }

  async getPermissionById(req, res) {
    try {
      const { id } = req.params;

      const [permissions] = await pool.query(
        'SELECT * FROM permissions WHERE id = ?',
        [id]
      );

      if (permissions.length === 0) {
        return res.status(404).json({
          success: false,
          message: '权限不存在'
        });
      }

      res.json({
        success: true,
        data: permissions[0]
      });
    } catch (error) {
      logger.error('获取权限详情失败:', error);
      res.status(500).json({
        success: false,
        message: '获取权限详情失败'
      });
    }
  }

  async createPermission(req, res) {
    try {
      const { permissionName, permissionCode, permissionType, parentId, path, component, icon, sortOrder, description, status = 1 } = req.body;

      const [existingPermissions] = await pool.query(
        'SELECT id FROM permissions WHERE permission_code = ?',
        [permissionCode]
      );

      if (existingPermissions.length > 0) {
        return res.status(400).json({
          success: false,
          message: '权限编码已存在'
        });
      }

      const [result] = await pool.query(
        'INSERT INTO permissions (permission_name, permission_code, permission_type, parent_id, path, component, icon, sort_order, description, status) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?)',
        [permissionName, permissionCode, permissionType, parentId || 0, path, component, icon, sortOrder || 0, description, status]
      );

      logger.info(`创建权限成功: ${permissionName}, 操作人: ${req.user.username}`);

      res.status(201).json({
        success: true,
        message: '创建权限成功',
        data: {
          id: result.insertId,
          permissionName
        }
      });
    } catch (error) {
      logger.error('创建权限失败:', error);
      res.status(500).json({
        success: false,
        message: '创建权限失败'
      });
    }
  }

  async updatePermission(req, res) {
    try {
      const { id } = req.params;
      const { permissionName, permissionType, parentId, path, component, icon, sortOrder, description, status } = req.body;

      const [permissions] = await pool.query('SELECT id FROM permissions WHERE id = ?', [id]);

      if (permissions.length === 0) {
        return res.status(404).json({
          success: false,
          message: '权限不存在'
        });
      }

      await pool.query(
        'UPDATE permissions SET permission_name = ?, permission_type = ?, parent_id = ?, path = ?, component = ?, icon = ?, sort_order = ?, description = ?, status = ? WHERE id = ?',
        [permissionName, permissionType, parentId || 0, path, component, icon, sortOrder || 0, description, status, id]
      );

      logger.info(`更新权限成功: ID=${id}, 操作人: ${req.user.username}`);

      res.json({
        success: true,
        message: '更新权限成功'
      });
    } catch (error) {
      logger.error('更新权限失败:', error);
      res.status(500).json({
        success: false,
        message: '更新权限失败'
      });
    }
  }

  async deletePermission(req, res) {
    try {
      const { id } = req.params;

      const [children] = await pool.query('SELECT COUNT(*) as count FROM permissions WHERE parent_id = ?', [id]);

      if (children[0].count > 0) {
        return res.status(400).json({
          success: false,
          message: '该权限下还有子权限，无法删除'
        });
      }

      const [permissions] = await pool.query('SELECT id FROM permissions WHERE id = ?', [id]);

      if (permissions.length === 0) {
        return res.status(404).json({
          success: false,
          message: '权限不存在'
        });
      }

      await pool.query('DELETE FROM permissions WHERE id = ?', [id]);

      logger.info(`删除权限成功: ID=${id}, 操作人: ${req.user.username}`);

      res.json({
        success: true,
        message: '删除权限成功'
      });
    } catch (error) {
      logger.error('删除权限失败:', error);
      res.status(500).json({
        success: false,
        message: '删除权限失败'
      });
    }
  }
}

module.exports = new PermissionController();

File: code/Api/controllers/permissionController.go
package controllers

import (
	"net/http"
	"superhoneypotguard/database"
	"superhoneypotguard/middleware"
	"superhoneypotguard/models"
	"superhoneypotguard/utils"
	"strconv"

	"github.com/gin-gonic/gin"
)

type PermissionController struct{}

func NewPermissionController() *PermissionController {
	return &PermissionController{}
}

func (ctrl *PermissionController) GetTree(c *gin.Context) {
	var permissions []models.Permission
	database.DB.Order("sort_order ASC, id ASC").Find(&permissions)

	tree := buildTree(permissions, 0)

	utils.SuccessResponse(c, tree)
}

func buildTree(permissions []models.Permission, parentId int) []models.Permission {
	var result []models.Permission
	for _, p := range permissions {
		if p.ParentID == parentId {
			p.Children = buildTree(permissions, p.ID)
			result = append(result, p)
		}
	}
	return result
}

func (ctrl *PermissionController) GetAll(c *gin.Context) {
	var permissions []models.Permission
	database.DB.Where("status = 1").Order("sort_order ASC, id ASC").Find(&permissions)

	utils.SuccessResponse(c, permissions)
}

func (ctrl *PermissionController) GetById(c *gin.Context) {
	id := c.Param("id")

	var permission models.Permission
	if err := database.DB.Where("id = ?", id).First(&permission).Error; err != nil {
		utils.ErrorResponse(c, http.StatusNotFound, "权限不存在")
		return
	}

	utils.SuccessResponse(c, permission)
}

func (ctrl *PermissionController) Create(c *gin.Context) {
	var req models.CreatePermissionRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		utils.ErrorResponse(c, http.StatusBadRequest, "参数验证失败")
		return
	}

	var count int64
	database.DB.Model(&models.Permission{}).Where("permission_code = ?", req.PermissionCode).Count(&count)
	if count > 0 {
		utils.ErrorResponse(c, http.StatusBadRequest, "权限编码已存在")
		return
	}

	parentId := 0
	if req.ParentID != nil {
		parentId = *req.ParentID
	}

	sortOrder := 0
	if req.SortOrder != nil {
		sortOrder = *req.SortOrder
	}

	status := 1
	if req.Status != nil {
		status = *req.Status
	}

	permission := models.Permission{
		PermissionName: req.PermissionName,
		PermissionCode: req.PermissionCode,
		PermissionType: req.PermissionType,
		ParentID:       parentId,
		Path:           req.Path,
		Component:      req.Component,
		Icon:           req.Icon,
		SortOrder:      sortOrder,
		Description:    req.Description,
		Status:         status,
	}

	if err := database.DB.Create(&permission).Error; err != nil {
		utils.ErrorResponse(c, http.StatusInternalServerError, "创建权限失败")
		return
	}

	utils.SuccessResponse(c, gin.H{
		"id":             permission.ID,
		"permissionName": permission.PermissionName,
	})
}

func (ctrl *PermissionController) Update(c *gin.Context) {
	id := c.Param("id")

	var permission models.Permission
	if err := database.DB.Where("id = ?", id).First(&permission).Error; err != nil {
		utils.ErrorResponse(c, http.StatusNotFound, "权限不存在")
		return
	}

	var req models.UpdatePermissionRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		utils.ErrorResponse(c, http.StatusBadRequest, "参数验证失败")
		return
	}

	parentId := 0
	if req.ParentID != nil {
		parentId = *req.ParentID
	}

	sortOrder := 0
	if req.SortOrder != nil {
		sortOrder = *req.SortOrder
	}

	status := 1
	if req.Status != nil {
		status = *req.Status
	}

	updates := map[string]interface{}{
		"permission_name": req.PermissionName,
		"permission_type": req.PermissionType,
		"parent_id":       parentId,
		"path":            req.Path,
		"component":       req.Component,
		"icon":            req.Icon,
		"sort_order":      sortOrder,
		"description":     req.Description,
		"status":          status,
	}

	if err := database.DB.Model(&permission).Updates(updates).Error; err != nil {
		utils.ErrorResponse(c, http.StatusInternalServerError, "更新权限失败")
		return
	}

	utils.SuccessResponse(c, nil)
}

func (ctrl *PermissionController) Delete(c *gin.Context) {
	id := c.Param("id")

	var count int64
	database.DB.Model(&models.Permission{}).Where("parent_id = ?", id).Count(&count)
	if count > 0 {
		utils.ErrorResponse(c, http.StatusBadRequest, "该权限下还有子权限，无法删除")
		return
	}

	var permission models.Permission
	if err := database.DB.Where("id = ?", id).First(&permission).Error; err != nil {
		utils.ErrorResponse(c, http.StatusNotFound, "权限不存在")
		return
	}

	if err := database.DB.Delete(&permission).Error; err != nil {
		utils.ErrorResponse(c, http.StatusInternalServerError, "删除权限失败")
		return
	}

	utils.SuccessResponse(c, nil)
}

func parseInt(s string) int {
	result, _ := strconv.Atoi(s)
	return result
}
